interpreter/prune.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interpreter
+
+import (
+	"github.com/golang/protobuf/ptypes/struct"
+	"github.com/google/cel-go/common/operators"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	expr "github.com/google/cel-spec/proto/v1/syntax"
+)
+
+type exprPruner struct {
+	expr    *expr.Expr
+	program Program
+	state   EvalState
+}
+
+// Prunes the given expression based on the given EvalState. Modifies the expression in place.
+// A typical use of this interface would be:
+// 1) Evaluate expr with some unknowns,
+// 2) If result is unknown:
+//   a) Maybe clone the expr(typically would be necessary only for the first iteration) and PruneExpr
+//   b) Goto 1
+// Functional call results which are known would be effectively cached across iterations.
+func PruneExpr(expr *expr.Expr, program Program, state EvalState) {
```

Comment:
Taking a step back, the `WalkExpr()` is a special case prune which omits constant values from the generated instruction set. It should be modified to honor previously set state in order to skip unnecessary instruction generation. Right now, `WalkExpr()` does too much. We should probably have a flow that looks as follows:

```go
evalState := NewMutableEvalState()
MapCommonSubexpressions(e *expr.Expr, state MutableEvalState)
FoldConstants(e *expr.Epxr, state MutableEvalState)
program := GenerateInstructions(e *expr.Expr, state EvalState)
```
Optionally a `GenerateAst(e *expr.Expr, state EvalState)` function could be added to output a new AST.

Thoughts?

Reply:
I agree with the overall thought that generating instructions/a new ast should be driven off from evalstate. I renamed PruneExpr to PruneAst. We can have finding common subexpression decoupled from generating instructions. I left this as a todo. GenerateInstructions could do something versy similar to PruneAst to avoid generating instructions, but I am guessing common use case would be to prune the ast so it can be cached. I removed the ConstantFold interface. As we discussed, users can bring together several tools and do the constant folding their way. I listed a typical uses in PruneAst interface.